components/renku_data_services/session/db.py

```diff
@@ -299,29 +341,89 @@ async def update_launcher(
-                launcher.container_image = None
-            if launcher.environment_kind == EnvironmentKind.container_image:
-                launcher.environment = None
-
-            launcher_model = launcher.dump()
-            models.SessionLauncher.model_validate(launcher_model)
+            env_payload: dict = kwargs.get("environment", {})
+            if len(env_payload.keys()) == 1 and "id" in env_payload and isinstance(env_payload["id"], str):
+                # The environment ID is being changed or set
+                old_environment = launcher.environment
+                new_environment_id = ULID.from_str(env_payload["id"])
+                res_env = await session.scalars(
+                    select(schemas.EnvironmentORM).where(schemas.EnvironmentORM.id == new_environment_id)
+                )
+                new_environment = res_env.one_or_none()
+                if new_environment is None:
+                    raise errors.MissingResourceError(
+                        message=f"Session environment with id '{new_environment_id}' does not exist or "
+                        "you do not have access to it."
+                    )
+                if new_environment.environment_kind != models.EnvironmentKind.GLOBAL:
+                    raise errors.ValidationError(
+                        message="Cannot set the environment for a launcher to an existing environment if that "
+                        "existing environment is not global",
+                        quiet=True,
+                    )
+                launcher.environment_id = new_environment_id
+                launcher.environment = new_environment
+                if old_environment.environment_kind == models.EnvironmentKind.CUSTOM:
+                    # A custom environment exists but it is being updated to a global one
+                    # We remove the custom environment to avoid accumulating custom environments that are not associated
+                    # with any launchers.
+                    await session.delete(old_environment)
+            else:
+                # Fields other than the environment ID are being updated
+                if launcher.environment.environment_kind == models.EnvironmentKind.GLOBAL:
+                    # A global environment is being replaced with a custom one
+                    if env_payload.get("environment_kind") == models.EnvironmentKind.GLOBAL:
+                        raise errors.ValidationError(
+                            message="When one global environment is being replaced with another in a "
+                            "launcher only the new global environment ID should be specfied",
+                            quiet=True,
+                        )
+                    env_payload["environment_kind"] = models.EnvironmentKind.CUSTOM.value
+                    env_payload_valid = apispec.EnvironmentPostInLauncher.model_validate(env_payload)
```

Comment:
thought: to be honest this sucks a bit, it is a lot of business type logic in the database repo. But in my PR that I based on top of yours, I tried to come up with a cleaner solution (where the logic on if we need to create a new environment, switch from global to global or not, etc. is handled in a core.py and we don't need to depend on the apispec outside of blueprints) but I could not come up with a solution that's any cleaner. 
The only thing that came to mind is not allowing posting an environment body here at all, only an ID, and requiring clients to creates environments by posting to an environment endpoint first before patching the launcher. But that's not as nice from the UX perspective of clients.

This method here is rather long though and I wonder if it should be split up to deal with e.g. environments and resource classes separately.

Reply:
Now it is cleaned up. Check the last commit in this PR.